rflx/sessionfile.py

```diff
@@ -0,0 +1,272 @@
+            if decl.identifier in result:
+                self.error.append(
+                    f"conflicting type {decl.identifier}", Subsystem.SESSION, Severity.ERROR,
+                )
+            result[decl.identifier] = decl
+        self.error.propagate()
+
+    def __parse_channels(self, doc: Dict[str, Any], result: Dict[ID, Declaration]) -> None:
+        if "channels" not in doc:
+            return
+        for index, f in enumerate(doc["channels"]):
+            if "name" not in f:
+                self.error.append(
+                    f"channel {index} has no name", Subsystem.SESSION, Severity.ERROR,
+                )
+                continue
+            if "mode" not in f:
+                self.error.append(
+                    f"channel {f['name']} has no mode", Subsystem.SESSION, Severity.ERROR,
+                )
+                continue
+            modes = {
+                "Read": {"read": True, "write": False},
+                "Write": {"read": False, "write": True},
+                "Read_Write": {"read": True, "write": True},
+            }
+            if ID(f["name"]) in result:
+                name = ID(f["name"])
+                self.error.append(
+                    f'conflicting channel "{name}"', Subsystem.SESSION, Severity.ERROR,
+                )
+            mode = f["mode"]
+            try:
+                result[ID(f["name"])] = ChannelDeclaration(
+                    f["name"], modes[mode]["read"], modes[mode]["write"]
+                )
+            except KeyError:
+                self.error.append(
+                    f"channel {f['name']} has invalid mode {f['mode']}",
+                    Subsystem.SESSION,
+                    Severity.ERROR,
+                )
+                continue
+        self.error.propagate()
+
+    def __parse_renames(self, doc: Dict[str, Any], result: Dict[ID, Declaration]) -> None:
+        if "renames" not in doc:
+            return
+        for index, f in enumerate(doc["renames"]):
+            try:
+                decl = declaration(f)
+            except RecordFluxError as e:
+                self.error.append(
+                    f"error parsing renames declaration {index}", Subsystem.SESSION, Severity.ERROR,
+                )
+                self.error.extend(e)
+                continue
+            if decl.identifier in result:
+                self.error.append(
+                    f"conflicting renames {decl.identifier}", Subsystem.SESSION, Severity.ERROR,
+                )
+            result[decl.identifier] = decl
+        self.error.propagate()
+
+    def __parse_declarations(self, doc: Dict[str, Any]) -> Dict[ID, Declaration]:
+        result: Dict[ID, Declaration] = {}
+        self.__parse_functions(doc, result)
+        self.__parse_variables(doc, result)
+        self.__parse_types(doc, result)
+        self.__parse_channels(doc, result)
+        self.__parse_renames(doc, result)
+        return result
```

Comment:
Wouldn't it be more efficient to iterate over `doc` and call the appropriate function for each key instead of testing inside each function if it is relevant?

Reply:
`sessionfile` is a temporary solution to support the YAML-based session specification. We will remove the YAML-support before the next release (cf. #398). So there is no need to improve its implementation.